### ISSUE TITLE
EagerRelationships should be pageable

### ISSUE THREAD

**Issue by agoncal** [jhipster/generator-jhipster#6791]
##### **Overview of the issue**

When there is a ManyToMany relationship, JHipster generates `findAllWithEagerRelationships` methods in the repositories. These methods should be pageable.

##### **Motivation for or Use Case**

Imagine that you have a ManyToMany relationship between a `SponsorAgreement` and a `Contact`. JHipster will generate the following code:

```
public interface SponsorAgreementRepository extends JpaRepository<SponsorAgreement, Long> {
    @Query("select distinct sponsor_agreement from SponsorAgreement sponsor_agreement left join fetch sponsor_agreement.contactsOthers")
    List<SponsorAgreement> findAllWithEagerRelationships();
}
```
The problem is that the method `findAllWithEagerRelationships` returns a `List` and not a `Page`. Instead, it should generated as follow:

```
public interface SponsorAgreementRepository extends JpaRepository<SponsorAgreement, Long>, JpaSpecificationExecutor<SponsorAgreement> {
    @Query(value = "select distinct sponsor_agreement from SponsorAgreement sponsor_agreement left join fetch sponsor_agreement.contactsOthers",
    countQuery = "select count(distinct sponsor_agreement) from SponsorAgreement sponsor_agreement")
    Page<SponsorAgreement> findAllWithEagerRelationships(Pageable pageable);
}
```

Notice the `countQuery`. Spring Data [needs this to be able to count the number of items](https://codingexplained.com/coding/java/spring-framework/fetch-query-not-working-spring-data-jpa-pageable) and calculate the number of pages.

BTW, the methods in the repository are not used (not even tested) when you use pagination.

##### **JHipster Version(s)**

```
featurepageable@0.0.0 /Users/agoncal/Documents/Code/Temp/jhipster/featurepageable
└── generator-jhipster@4.11.1 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "org.agoncal.featurepageable"
    },
    "jhipsterVersion": "4.11.1",
    "baseName": "featurepageable",
    "packageName": "org.agoncal.featurepageable",
    "packageFolder": "org/agoncal/featurepageable",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Memory",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "replaced-by-jhipster-info",
    "clientFramework": "angularX",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": false
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Contact (contact) {
  firstName String,
  lastName String
}
entity SponsorAgreement (sponsor_agreement) {
  vatRate Float,
  vatAmount Float,
  total Float
}

relationship ManyToMany {
  SponsorAgreement{contactsOthers(lastName)} to Contact
}

paginate Contact, SponsorAgreement with pagination
service Contact, SponsorAgreement with serviceClass

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_152"
Java(TM) SE Runtime Environment (build 1.8.0_152-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.152-b16, mixed mode)

git version 2.15.0

node: v8.9.0

npm: 5.5.1

- [X] Checking this box is mandatory (this is just to show you read everything)

**Comment by ruddell** [comment]
Fixed in #6805

**Comment by micobarac** [comment]
I have a problem with latest JHipster and it's generated repository code:

Inside Role class I have a @ManyToMany relationship with Action:

```java
    @ManyToMany
    @Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
    @JoinTable(name = "role_action",
               joinColumns = @JoinColumn(name = "role_id", referencedColumnName = "id"),
               inverseJoinColumns = @JoinColumn(name = "action_id", referencedColumnName = "id"))
    private Set<Action> actions = new HashSet<>();
```

JHipster generated this RoleRepository query:

```java
    @Query(value = "select distinct role from Role role left join fetch role.actions",
        countQuery = "select count(distinct role) from Role role")
    Page<Role> findAllWithEagerRelationships(Pageable pageable);
```

This query throws an exception:

`firstResult/maxResults specified with collection fetch. In memory pagination was about to be applied. Failing because 'Fail on pagination over collection fetch' is enabled.; nested exception is org.hibernate.HibernateException: firstResult/maxResults specified with collection fetch. In memory pagination was about to be applied. Failing because 'Fail on pagination over collection fetch' is enabled.`
